FOCUS #185: Key-Value Format (#219)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/attributes/key_value_format.md

# Key-Value Format

> This is the specified format for providing data in the form of key-value pairs, which are also known as name-value pairs, attribute-value pairs, and field-value pairs in other contexts. 

Reviewer: For consistency, you might consider adding a sentence like we do for other format attributes where we explain why this would be helpful to consumers.

This format is often used in place of enumerating data which would be inherently sparse and/or without predetermined keys (attributes) into separate columns. To faciliate parsing, this format does not support nesting.

> 

Reviewer 2: From your initial paragraphs, tried to a bit less technical.  Feel free to judge as you wish ...

The Key-Value Format specifies the format for providing data as key-value pairs, also known as name-value pairs, attribute-value pairs, and field-value pairs in other contexts.

The Key-Value Format mirrors JSON's ECMA-404 with a handful or extra requirements and in most cases, more efficiently consolidates column-based data than the CSV format.

Reviewer 2: Also removed this line "To faciliate parsing, this format does not support nesting." because it seemed duplicated in the requirements below

Reviewer: @cnharris10 Not seeing where this is mentioned in requirements

Reviewer 2: Referring to this line, `* Values MUST NOT be an object nor an array.`.  If this is true, nesting is not possible.

> 

Reviewer: To be consistent with the other attributes, we should add a normative line here that says MUST meet all requirements listed below (in the requirements section)

> All key-value related columns defined in the FOCUS specification MUST follow the key-value formatting requirements listed below. These columns SHALL NOT contain nested elements.

Reviewer: Is the last sentence needed here? It's one of the requirements listed in the requirements already.

Reviewer 3: No reason to duplicate

Author: It was intended to make this more approachable, but the common opinion is that this redundant comment doesn't add value.

## Attribute ID

KeyValueFormat

## Attribute Name

> Key-Value Format

Reviewer: Should it be KeyValue Format? 

I ask because we also have Date/Time Format and it feels like it would be better to not have - or /. Otherwise we should pick one Applies here and for the top of the page header (see [here](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/working_draft/specification/attributes/datetime_format.md))

Reviewer 3: I like Keyvalue, DateTime....

Reviewer 4: Wikipedia uses [key-value](https://en.wikipedia.org/wiki/Key%E2%80%93value_database) and [date/time](https://en.wikipedia.org/wiki/ISO_8601#EDTF) just to keep us on our toes 🙃 I believe "DateTime" is slightly different in that it refers to a datatype in a programming language. I vaguely recall looking that up when we went thru that one.

Author: To me, the Attribute Name is written in plain English as you would when using it in an email and Attribute ID is written in the way you'd use it in a system or reference it in a slack message with technical teams.

In English, it's very atypical to have a compound word written in Pascal case.

For now, we can leave it as key-value since we have it written as date/time in the working draft. If we want to change this, we should do both (and any similar examples we find) in the same PR.

## Description

Rules and formatting requirements for columns appearing in billing data which convey data as key-value pairs.

## Requirements

> * Columns subject to Key-Value Format MUST be a serialized JSON string, consistent with the ECMA-404 definition of an object.

Reviewer: ```suggestion
* Columns of type Key-Value Format MUST contain serialized JSON string, consistent with the ECMA-404 definition of an object.
```

* Keys MUST be unique within an object.

> * Values MAY be a number, string, `true`, `false`, or `null`.

Reviewer 2: Another option: `Values MUST be either a ...`

> * Values MAY be a number, string, `true`, `false`, or `null`.

Reviewer: Is it clear that this is referring to the Key and Value of each KV pair defined in this column? Line 23 for example could get confused with the overall value of the cell? (I could be misreading that)

> * Key-Value Format columns MUST contain a serialized JSON string, consistent with the ECMA-404 definition of an object.

Reviewer 5: Do we want to have an opinion on case sensitivity? As it stands this is permitting "keynull" and "KeYNuLl". Some CSP's have case sensitive tags, some CSP's don't.

Reviewer 5: never mind. it is probably use case dependent (e.g. this is a discussion better left for `Tags`)

>   * Values MUST NOT be an object nor an array.

Reviewer 2: ```suggestion
 * Values MUST NOT be an object nor an array.
```
